### ISSUE TITLE
Feature request: Support configuration or new task that doesn't package a Java app, just specified paths.

### ISSUE THREAD

**Issue by nhoughto** [GoogleContainerTools/jib#3487]
Jib and the associated Gradle plugin look to solve a number of problems i've come across, and whilst the jib-cli is a great escape hatch to supporting any kind of app/packaging needs the gradle plugin specifically is a great help ergonomically, it avoids having to create jib files and makes it easy to programmatically change jib config per module/project. So I want to use jib and jib gradle plugin, but not all of my containers are Java and even those that are Java I don't really want to package the app the way Jib wants to out of the box (exploded or packaged modes).

A good way to solve both of these issues would be to tweak the jib gradle plugin to not always assume that a Java app is being packaged, make it an optional addition (defaulted to on for backwards compat) with the disabled state just packaging all the files specified in the `extraDirectories` section. This way I can use Jib and the gradle plugin, avoid jib files and keep simple per module config AND use the same system for both Java apps and nginx and stubs.. etc all of the variety of apps. A slightly different more gradle-aware approach would be to support Gradle Distributions/Applications, whereby Jib aims to package a Gradle application (essentially just a zip/tar constructed by Gradle - https://docs.gradle.org/current/userguide/application_plugin.html) and what is in that application is controlled by Gradle thus breaking the Java app assumption/dependency.

My current workaround is to apply `java-library' plugin to all my gradle projects so Jib does all the packaging of essentially nothing, just to avoid it failing because the 'jar' task doesn't exist! Quite the silly workaround =)

**Comment by suztomo** [comment]
Is your project publicly available? I'm curious about the concrete implementation of your workaround.

I'm afraid that expanding the Jib Gradle plugin to other languages is not in our current roadmap. Did you check whether buildpacks helps your use case?

**Comment by nhoughto** [comment]
Not public but basic gradle script to workaround it looks like this, it still put a few unused files in the container (not ideal) but they can be ignored I suppose:

```
apply plugin: 'java-library' //shouldnt need this
apply plugin: com.google.cloud.tools.jib.gradle.JibPlugin

configurations {
    dummyJib
}

//not configurable via gradle, be nice to configure via 'jib' extension like other things
System.setProperty("jib.baseImageCache", "${rootProject.projectDir}/.gradle/jib-base-cache")
System.setProperty("jib.applicationCache", "${rootProject.projectDir}/.gradle/jib-application-cache")
System.setProperty("jib.console", "plain")

jib {
    from {
        image = containerFromImage
    }
    to {
        image = containerToImage
    }
    container {
        workingDirectory = "/app"
        entrypoint = "/app/bin/service"
        format = "OCI"
        mainClass = "Dummy" //avoids annoying warning, not used anyway
    }
    extraDirectories {
        paths {
            path {
                from = file("${buildDir}/distributions-unpacked/")
                into = "/app"
            }
        }
    }
    containerizingMode = "packaged"
    configurationName = "dummyJib"
}

task distDir (type: Copy) {
    with distributions.main.contents
    into file("${buildDir}/distributions-unpacked/")
}
```

Use `distDir` to workaround the difference between jib and Dockerfile with `ADD` auto-unpacking tar files.

Haven't looked at buildpacks as that seems like a replacement for a bunch of the existing gradle stuff already has, like the Gradle application plugin and i'm not going to tear it all out just to build a container =).

**Comment by chanseokoh** [comment]
> it still put a few unused files in the container (not ideal)

Just so you know, you can use the [Jib Layer-Filter Extension](https://github.com/GoogleContainerTools/jib-extensions/tree/master/first-party/jib-layer-filter-extension-gradle) to fine-tune files in image layers, including emove files.

**Comment by chanseokoh** [comment]
Also take a look at [these Stack Overflow answers](https://stackoverflow.com/q/66331028/1701388), if it helps at all.

**Comment by nhoughto** [comment]
Yep 👍 i spotted the layer filter, intend to use it to do exactly that, and obv do some layering.

Is there any doco on best practice for image layering in a multi-module project running in parallel? Intent of layering is to re-use layers, but if all the modules are building all the layers at the same time I expect this would lose some of the benefit? With an old world `docker build ..` the docker daemon is the central point that de-dupes work like pulling remote images etc across docker build jobs, I imagine `jib` doesnt/cant do that as there are N jib processes running one for each parallel executing gradle task?

**Comment by chanseokoh** [comment]
Jib does cache base image layers and application layers.

The base image layer cache is shared by all Jib tools (Maven and Gradle plugins and Jib CLI). So, as long as base image layers are cached once, they will be reused by any number of concurrent threads or processes running Jib tools.

OTOH, the application layer cache is by default local to each project (`build/jib-cache` for Gradle and `target/jib-cache` for Maven). Usually there's no value to share layers between projects, since each project is unique and it's rare that two or more projects build the exact same layers. However, if you have multiple projects that at least have some layers that are built identically, I think it's not a bad idea to share them by setting their application cache directory to a global location. But usually application layers are small and different most of the time. Check out the [`jib.applicationCache`](https://github.com/GoogleContainerTools/jib/blob/master/jib-gradle-plugin/README.md#system-properties) system property.

Lastly, I think you already know this, but just in case: in order for two layers (which are simply `.tar` files) to be identical, they should be equal byte-to-byte, resulting in the same digest checksum. Even having a different timestamp for a file makes it different. Good news is that, Jib wipes out highly volatile and system-dependent information like file timestamps, permissions, etc., so it results in strong reproducibility and cacheability, saving storage, network bandwidth, and time.

**Comment by nhoughto** [comment]
yep makes sense, ok so coordinating concurrent jib builds to share layers will need to be done outside of jib/in gradle itself.
My scenario is building 20~ gradle application distributions, they are all java services that share most 3rd party dependencies, this should be easy to split out into its own layer that could be shared. Seems like i've got all I need to make that happen, thanks.
